src/js/snp_attestation.cpp

```diff
@@ -0,0 +1,299 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+
+#include "ccf/js_plugin.h"
+#include "ccf/js_snp_attestation_plugin.h"
+#include "ccf/pal/attestation.h"
+#include "ccf/version.h"
+#include "js/wrap.h"
+#include "node/uvm_endorsements.h"
+
+#include <algorithm>
+#include <quickjs/quickjs.h>
+#include <regex>
+#include <vector>
+
+namespace ccf::js
+{
+#pragma clang diagnostic push
+
+  static JSValue make_js_tcb_version(JSContext* ctx, pal::snp::TcbVersion tcb)
+  {
+    auto js_tcb = JS_NewObject(ctx);
+    JS_SetPropertyStr(
+      ctx, js_tcb, "boot_loader", JS_NewUint32(ctx, tcb.boot_loader));
+    JS_SetPropertyStr(ctx, js_tcb, "tee", JS_NewUint32(ctx, tcb.tee));
+    JS_SetPropertyStr(ctx, js_tcb, "snp", JS_NewUint32(ctx, tcb.snp));
+    JS_SetPropertyStr(
+      ctx, js_tcb, "microcode", JS_NewUint32(ctx, tcb.microcode));
+    return js_tcb;
+  }
+
+  static JSValue JS_NewArrayBuffer2(
+    JSContext* ctx, std::span<const uint8_t> data)
+  {
+    return JS_NewArrayBufferCopy(ctx, data.data(), data.size());
+  }
+
+  static JSValue js_verify_snp_attestation(
+    JSContext* ctx, JSValueConst, int argc, JSValueConst* argv)
+  {
+    if (argc != 3 && argc != 4)
+      return JS_ThrowTypeError(
+        ctx, "Passed %d arguments, but expected 3 or 4", argc);
+    js::Context& jsctx = *(js::Context*)JS_GetContextOpaque(ctx);
+
+    size_t evidence_size;
+    uint8_t* evidence = JS_GetArrayBuffer(ctx, &evidence_size, argv[0]);
+    if (!evidence)
+    {
+      js::js_dump_error(ctx);
+      return JS_EXCEPTION;
+    }
+
+    size_t endorsements_size;
+    uint8_t* endorsements = JS_GetArrayBuffer(ctx, &endorsements_size, argv[1]);
+    if (!endorsements)
+    {
+      js::js_dump_error(ctx);
+      return JS_EXCEPTION;
+    }
+
+    size_t uvm_endorsements_size;
+    uint8_t* uvm_endorsements_array =
+      JS_GetArrayBuffer(ctx, &uvm_endorsements_size, argv[2]);
+    if (!uvm_endorsements_array)
+    {
+      js::js_dump_error(ctx);
+      return JS_EXCEPTION;
+    }
+    auto uvm_endorsements = std::vector<uint8_t>(
+      uvm_endorsements_array, uvm_endorsements_array + uvm_endorsements_size);
+
+    QuoteInfo quote_info = {};
+    quote_info.format = QuoteFormat::amd_sev_snp_v1;
+    quote_info.quote = std::vector<uint8_t>(evidence, evidence + evidence_size);
+    quote_info.endorsements =
+      std::vector<uint8_t>(endorsements, endorsements + endorsements_size);
+    if (argc == 4)
+    {
+      quote_info.endorsed_tcb = jsctx.to_str(argv[3]);
+    }
+
+    pal::PlatformAttestationMeasurement measurement = {};
+    pal::PlatformAttestationReportData report_data = {};
+    UVMEndorsements parsed_uvm_endorsements;
+
+    try
+    {
+      pal::verify_snp_attestation_report(quote_info, measurement, report_data);
+      parsed_uvm_endorsements =
```

Comment:
This mandates uvm endorsements in the current format, and does not allow the caller to pass a trusted platform measurement instead, or to validate it against another form of endorsement.

Reply:
Confirmed with Kapil that manual checks need to be supported, so this step needs to be made optional when no UVM endorsements are passed, and the caller will then check `measurement` separately.

Reply:
https://github.com/microsoft/CCF/pull/5653/commits/b317000c0b47e8fbe2c61cb53ac3d11051a776b8